### PR TITLE
[Migrator] Add `migrate-sass-transition` migration for transition durations & easing

### PR DESCRIPTION
**Pull request by jesstelford**
_NOTE: This leverages the functionality in #7543 to get some sweet sweet clean code and juicy verbose output!_

### WHY are these changes introduced?

ref https://github.com/Shopify/polaris/issues/7213

This PR covers the `transition`, `transition-duration`, `transition-delay`, and `transition-timing-function` properties, their various complex uses and a couple of edge cases as seen in the wild.

### WHAT is this pull request doing?

Example migrations (taken from the test files):

#### `duration`

* `transition-duration: 100ms;`
    ↳ `transition-duration: var(--p-duration-100);`
* `transition-duration: legacy-polaris-v8.duration('slow');`
    ↳ `transition-duration: var(--p-duration-300);`
* `transition: opacity 100ms linear;`
    ↳ `transition: opacity var(--p-duration-100) linear;`
* `transition: opacity legacy-polaris-v8.duration('slow') linear;`
    ↳ `transition: opacity var(--p-duration-300) linear;`
* `transition: opacity 100ms linear, left 100ms linear;`
    ↳ `transition: opacity var(--p-duration-100) linear, left var(--p-duration-100) linear;`
* `transition: opacity legacy-polaris-v8.duration(slow) linear, left legacy-polaris-v8.duration(slow) linear;`
    ↳ `transition: opacity var(--p-duration-300) linear, left var(--p-duration-300) linear;`
* `transition: opacity legacy-polaris-v8.duration(slower) linear legacy-polaris-v8.duration(fast);`
    ↳ `transition: opacity var(--p-duration-400) linear var(--p-duration-100);`

#### `easing`

* `transition-timing-function: linear`
    ↳ `transition-timing-function: var(--p-linear)`
* `transition-timing-function: legacy-polaris-v8.easing('in')`
    ↳ `transition-timing-function: var(--p-ease-in);`
* `transition: opacity 300ms legacy-polaris-v8.easing('base');`
    ↳ `transition: opacity var(--p-duration-300) var(--p-ease);`

Will output various error messages for things it can't migrate:

* `transition-timing-function: legacy-polaris-v8.easing(anticipate);`
    ↳
    ```
    /* polaris-migrator: Unable to migrate the following expression. Please upgrade manually. The anticipate easing function is no longer available in Polaris. See https://polaris.shopify.com/tokens/motion for possible values. */
    transition-timing-function: legacy-polaris-v8.easing('anticipate');
    ```
* `transition-timing-function: cubic-bezier(0, 0, 1, 1);`
    ↳
    ```
    /* polaris-migrator: Unable to migrate the following expression. Please upgrade manually. Unexpected easing function 'cubic-bezier'. See https://polaris.shopify.com/tokens/motion for possible values. */
    transition-timing-function: cubic-bezier(0, 0, 1, 1);
    ```

---

## Notes

1. We consider `cubic-bezier()` / `linear()` / other built-in CSS easing functions as failures (inserting a comment when they're detected).
2. We consider use of variables (`$foo`) as failures (because we can't statically analyse what their value is).

### CODE REVIEW

**Review by gwyneplaine — APPROVED** [review]
lgtm

**Comment by gwyneplaine** [comment]
/stageit
woops wrong command

File: polaris-migrator/src/utilities/sass.ts [hunk 1]
```diff
@@ -249,6 +297,33 @@ export function toTransformablePx(value: string) {
...
+  // Zero is the only unitless dimension our duration transforms support
+  if (isUnitlessZero(dimension)) return true;
+
+  return transformableDurationUnits.includes(dimension.unit);
+}
+
+export function isPolarisVar(node: Node): boolean {
+  return (
+    isSassFunction('var', node) &&
+    (node.nodes?.[0]?.value ?? '').startsWith('--p-')
```
Comment (sam-b-rose) [inline polaris-migrator/src/utilities/sass.ts]: Maybe we can be extra sure this is a Polaris token by leveraging this [RegEx logic here](https://github.com/Shopify/polaris/pull/7375/files#diff-3ed201423f7ef97686e85c262e4c5adf310c70a0576e6c9c51a87a6887465620R242-R244)

**Review by sam-b-rose — APPROVED** [review]
Very thorough! I say we keep this migration _moving_ and motion forward 🚀 

We will likely want to split out the Sass function replacement with the tokenization of declaration values, but for now this is a safe start so we can start migrating in Shopify/web.

File: polaris-migrator/src/utilities/sass.ts (hunk 1 above)
Comment (jesstelford) [reply polaris-migrator/src/utilities/sass.ts]: We could, however that would mean we need to generate a gigantic regex that contains _all_ polaris tokens (since we're not only checking for `color`s as in that linked example).

Is there a reason the prefix check wouldn't work? Perhaps it returns false positives?

Comment (sam-b-rose) [reply polaris-migrator/src/utilities/sass.ts]: Just added safety is all. More concerned around other 3P apps using the migration and _possibly_ having custom properties prefixed with `--p-` as well. Low likelihood but possible 🤷  I'm happy with your current logic as well. My comment was more of a suggestion for added confidence that we are replacing the right tokens.

Comment (jesstelford) [reply polaris-migrator/src/utilities/sass.ts]: Ah, good point: I forgot about 3P devs! I've made that change 👍
